### PR TITLE
fix #14905: new recommended flag for performance/benchmarking: -d:nimDisableAssertMsgs

### LINKED ISSUE #14905: assert/doAssert makes unrelated code changes cause  1.4X performance differences, leading to wrong conclusions
## TLDR
use the upcoming `-d:nimDisableAssertMsgs` when benchmarking otherwise you may end up drawing wrong conclusions when deciding on whether algorithm1 or algorithm2 is better (which could include anything, including compiler flags etc): up to 1.4x perf difference caused by completely unrelated code changes such as adding an empty newline

## details
I was trying to debug a really weird performance issue where depending on unrelated code changes, some algorithm1 would sometimes be better than some algorithm2, and vice versa; by a performance factor of 1.15x to 1.4x; unrelated code changes could simply mean:
* adding an empty newline
* changing a variable name, etc.
* changing the test file name

After investigation, it turns out the culprit is assert+doAssert messages, more precisely:
* `assertImpl` generates ```failedAssertImpl(ploc & " `" & expr & "` " & msg)```
* `ploc` depends on file/line/col (which explains the above observations wrt adding empty newline or changing filename)
* `expr` depends on assert expression, eg: `doAssert fun() != 1` vs `doAssert fun2() != 1`, (which explains above observations wrt changing variable name)

## analysis: performance is bimodal but not monotonic wrt msg.len
* performance is not monotonic wrt msg.len: a code change that causes msg.len to be larger can lead to better or worse performance depending on the whole program and compiler options (eg `--passc:-flto --passl:-flto` passed or not)
* performance is typically bimodal: depending on msg length, you end up with performance equal to t1 or t2, in other words, `perf(len)` is a non-monotonic (could go up and down) function of len with 2 discrete values t1 < t2, where t2/t1 can be as large as 1.4x
* but in more complex examples (eg with M hotspots), we'd likely have 2^M modes, so for complex programs this may obscure the underlying issue

## explanation
* I believe this is due to C/C++ optimizer (in my case clang on OSX but other compilers likely have similar behavior) deciding to inline some piece of code or not depending on context, compiler flags (eg `--passc:-flto --passl:-flto`) and on length of static string constants, so that a variation of 1 in string length causes some inlining to happen or not.

### Example
```nim
when true: # gitissues D20200704T184212
  from times import cpuTime
  template benchmarkDisp*(name, n, body): untyped =
    let t = cpuTime()
    for i in 0..<n:
      body
    let t2 = cpuTime() - t
    echo (name, t2)

  template doAssert2(cond): untyped =
    doAssert cond # has bug
    # if not cond: quit(1) # removes bug

  proc bar2*[T](a: var openArray[T]): int =
    result = 0
    let n = len(a)
    for i in 0..<n:
      result += a[i]
      a[i] = result

  proc main =
    #[
    depending on n, you get different (stable) rankings results for:
    case1a, case1b, case1c, case1d

    using or not: --passc:-flto --passl:-flto
    ]#
    let n = 10
    #[
    eg: try:
    10,20,50,100,1000
    ]#

    let m = 100000000
    var s = newSeq[int](n)
    for i in 0..<n: s[i] = i*2
    const k = 1
    const k2asdfasfasfdasfd = 1
    const kasdfasdfasfdasfasfasfdasfasdfasfafdasdfasdfasdfasdfasfdasdfafdafdasdfasdfasdfasfdasfasfdasdfasdfasdfasdfasdfdfasdfasdfasfdasfasfdasdfasdfasdfasdfasdf = 1
    const k2as2 = 1 # bad
    const k2as21 = 1 # good
    benchmarkDisp("bar.alias", m):
      when defined case1a:
        doAssert2 bar2(s) != k

      when defined case1b:
        doAssert2 bar2(s) != k2as2

      when defined case1c:
        doAssert2 bar2(s) != kasdfasdfasfdasfasfasfdasfasdfasfafdasdfasdfasdfasdfasfdasdfafdafdasdfasdfasdfasfdasfasfdasdfasdfasdfasdfasdfdfasdfasdfasfdasfasfdasdfasdfasdfasdfasdf
      
      ## SKIP SKIP SKIP
      ## skip to line 100 (it matters for `assertImpl` since it adds 1 character in assert msgs)

      when defined case1d: # same as case1b but after line 100
        doAssert2 bar2(s) != k2as2

  main()

```

```
* make sure to insert blanks so that `skip to line 100` appears at line 100
* try this cmdline with 1 of: -d:case1a, -d:case1b, -d:case1c, -d:case1d
* with and without `--passc:-flto --passl:-flto`
* with different values of n (eg: 10,20,50,100,1000)
nim r -d:danger -d:case1a main.nim
```

### Current Output
* depending on parameters, you'll see performance changing in a bimodal way, so that for a fixed n and compiler options, -d:case1a, -d:case1b, -d:case1c, -d:case1d would give a runtime of just 2 values t1 or t2 (relatively stable across re-runs), with t2/t1 up to 1.4x depending on conditions.
* sometimes case1a wins, sometimes case2a wins, ... etc.
* `--passc:-flto --passl:-flto` sometimes helps, sometimes hurts
* eg, in some conditions, you'll get t1 = 0.40 and t2 = 0.27 => ~ 1.450 ratio

### Possible Solution
* add a `-d:nimDisableAssertMsgs` flag to disable messages (see upcoming PR)

### Additional Information
* devel 1.3.5 af27e6bdea63bbf66718193ec44bc61e745ded38

### PR DESCRIPTION
**Pull request by timotheecour**
* fix #14905 (make sure you read that carefully and understand it before reading this PR)
* `-d:nimDisableAssertMsgs` completely optimizes away the code that generates the assert/doAssert message: ```ploc & " `" & expr & "` " & msg```

## example1
```nim
# D20200704T184212
proc main[T](x, y: T) =
  proc bar(): string =
    echo "in bar" # wont be executed with -d:nimDisableAssertMsgs
    return "foo"
  doAssert x == y, $(x, y, bar())
main(12, 13)
```
* note that the assert/doAssert message generation generates code bloat (pressure on instruction cache) even if it's not executed at runtime, which can cause performance drop as well as the issues mentioned in #14905 where unrelated code changes would cause 1.4X performance differences, because it can affect C compiler's inlining decisions
* it would not be enough to drop `ploc` or `expr` or `msg`, as either of these would still affect code bloat (eg preventing some inlining to happen depending on context, where the context is hard/impossible to model)
* note that the argument "use assert instead of doAssert if you care about performance or in benchmarks" is invalid, because doAssert is still useful both in danger builds (eg to do a cheap but important check) or in benchmarks (eg to avoid optimizing away some code)

## example2: 11x speedup via `-d:nimDisableAssertMsgs`
speedup from `-d:nimDisableAssertMsgs` can be arbitrarily large, eg:

nim r -d:danger -d:case2 $timn_D/tests/nim/all/t11044.nim
with -d:nimDisableAssertMsgs: 0.539
without -d:nimDisableAssertMsgs: 6.60

note that in this particular simple example, `--passc:-flto --passl:-flto` would've given same result for both, but `--passc:-flto --passl:-flto` doesn't always help (eg, nim itself compiled with `--passc:-flto --passl:-flto -d:danger`  is slower than with `-d:danger`).

```nim
when true:
  # import timn/exp/benchutils2
  from times import cpuTime
  template benchmarkDisp*(name, n, body): untyped =
    let t = cpuTime()
    for i in 0..<n:
      body
    let t2 = cpuTime() - t
    echo (name, t2)

  proc bar(n: int): int =
    result = 0
    for i in 0..<n:
      result += (i*i) div n
      doAssert result != -3, $(result, i, n)
  proc main() =
    var tot = 0
    benchmarkDisp("bar", 10):
      tot += bar(1_00000000)
    echo tot
  main()
```

### CODE REVIEW

**Comment by juancarlospaco** [comment]
@timotheecour 
I thought something like this but for CI speed up, tons of assert there, have you tried making assert take varargs or openarray?. :+1:

**Comment by timotheecour** [comment]
> have you tried making assert take varargs or openarray?

this won't help, see analysis in #14905 ; the problem is this code: ```ploc & " `" & expr & "` " & msg```, which causes code bloat (eg puts pressure on instruction cache, or can prevent inlining, depending on context; and worst of all can lead to wrong performance conclusions that depends on unrelated context)

note that this is the case even when `msg = ""`, because of `ploc` and `expr` (as shown in #14905), so varargs/openArray will make no difference.

**Comment by juancarlospaco** [comment]
But I meant to remove all that code with `when defined():` from the root impl, not even empty string nor `"?"`, but I believe you.

**Comment by Varriount** [comment]
Aside from micro-benchmarks, was this an issue in  any larger program?

**Comment by timotheecour** [comment]
a main use case is getting more reliable relative performance for micro-benchmarks. Without this feature, you run the risk of optimizing for the wrong thing and deriving flawed conclusions, eg algo1 is faster than algo2 (and thus, picking the wrong algorithm in larger programs), whereas the opposite might be true.

**Comment by Varriount** [comment]
Why would one be using doAssert in a benchmark though?

**Comment by timotheecour** [comment]
for the same reason we have both assert and doAssert; not all conditions should be disabled:
* conditions that have low impact on performance or critical for security (low impact for checking the condition, after applying `-d:nimDisableAssertMsgs`)
* to make sure some logic doesn't get optimized away

note that due to aggressive inlining, doAssert can be in parts of code that seem "far" form hotspot and yet, without `-d:nimDisableAssertMsgs`, would affect the hotpsot because of code bloat

**Comment by Varriount** [comment]
I'm still uncomfortable with this - I feel like this addition is for a rather niche case, for testing conditions that are unstable at best. I'd almost prefer just a documentation change on doAssert warning about this, and possibly a static parameter enabling/disabling the message.

**Comment by timotheecour** [comment]
I don't see any downside; you don't have to use it if you don't want to; it doesn't add any complexity and has 0 impact if you don't use `-d: nimDisableAssertMsgs`; no existing code is affected and it's a tiny patch. I can reword the section `Recommended flags for highest performance and for benchmarking` to make it a mere suggestion when doing benchmarking instead of a recommendation if needed.

**Comment by Araq** [comment]
I too consider it an important problem to solve but the solution is rather inelegant and I wonder if we can do better.

**Comment by timotheecour** [comment]
> solution is rather inelegant and I wonder if we can do bette

inelegant because it uses a define? suggestions welcome..

I could introduce `--experimental:disableAssertMsgs` or introduce `--feature:enableAssertMsgs` (default:on), see below

## --feature:foo
this this isn't really an experimental feature, `--feature` might be more adequate, and avoids double negation. `--feature` would behave semantically (and could share implementation of) `--experimental`, so that it can be pushed/popped in a scope or set on cmdline. 

this keeps popping up and would be useful in other cases, eg https://github.com/nim-lang/Nim/pull/14904

## pros/conss of 3 common types of settings
One important aspect is avoid boilerplate when adding new features:
* minimum boilerplate is defined(foo); easiest, but has drawback of lack of name scoping and compiler doesn't help if you mistype
* medium boilerplate is `--experimental:foo` (or the proposed --feature:foo), which can be pushed/popped or passed on cmdline; it gives you type safety but can't be customized (eg `on|off|hints` would not work at least not unless we enhance this)
* high boilerplate is custom flags, eg --stylecheck:on|off|hints; it adds more boilerplate, can't be pushed/popped as easily (needs custom code), but is more flexible (eg can extend as was done wit --stylecheck:hints)

## execution lifetime vs scope lifetime
one subtlety is that `define` can be set/unset(via {.undef.}) but it's tied to VM execution lifetime, not to a scope; this is in fact a very useful property in some cases, eg when debugging, to enable compiler logging in some execution lifetime.

**Comment by Araq** [comment]
No, inelegant because maybe we can find a pattern so that ``if cond: prepareArgsForFormat()`` uses a minimal code size for ``prepareArgsForFormat``. But it looks like a hard problem.

**Comment by timotheecour** [comment]
there's an inherent tradeoff bw performance (IR cache), binary size, and debuggability, and as soon as you introduce runtime code in your hotspot frame, it can impact inlining by pressuring IR cache (even a single byte different in msg length can put you in a different regime, see analysis in https://github.com/nim-lang/Nim/issues/14905: `analysis: performance is bimodal but not monotonic wrt msg.len`

PTAL, I'm now distinguishing these 2 cases:
* -d:nimDisableAssertMsgs: 100% geared towards avoiding code bloat and minimizing binary size; it may also be useful for generating binaries with 0 debug info (eg to avoid leaking implementation details) and may save the day for performance/benchmarking/choosing between algo1 vs algo2. obvious caveat: 0 context on assert failure.
* new: -d:nimDisableAssertComputedMsgs: miminizes code bloat around the assert calls but can increase binary size; see implementation: it uses a 0-argument function call and discards the runtime message if it can't be computed at CT, but keeps the loc+expr(+ maybe msg if computable at CT); each assert location in source code gives rise to a separate function call.
 Since it's a separate function call and has 0 arguments, it doesn't affect hotspot much, but it generates a lot of "tiny functions", and may increase code size.
**Note that with example2 in top post, `-d:nimDisableAssertComputedMsgs` is as fast as `-d:nimDisableAssertMsgs` (but generates more useful asserts since it shows `loc+expr`; but not `msg` since that one can't be computed at CT in this example.**

## future work: encode file/line/col/(procname/sourceline etc) using a single `int` index, + a lookup table
this will decrease binary size significantly, and may improve performance in some cases.
the encoding is simply (uint16,uint16) for fileIndex+line(+col), and indexes into a lookup table (which is stored separately so won't affect hotspots, so long it's accessed rarely, as should be the case here.

Note that I need the same encoding for the 2nd part of https://github.com/nim-lang/Nim/pull/13630, which IIRC gave an extra 2X speedups for `--stacktrace:on` on top of the linkedlist=>array speedup

**Comment by alaviss** [comment]
Have you considered making `failedAssertImpl` taking `varargs[string]` and do the concat in there?

**Comment by alaviss** [comment]
I'm not a fan of how this is being implemented and instead would prefer a solution that doesn't require the use of even more defines to alter the behavior. We should do profiling and optimize the code so that it'd perform optimally without degrading the debugging experience.

CC @mratsim for his expertise in this sort of optimizations :)

**Comment by timotheecour** [comment]
> Have you considered making failedAssertImpl taking varargs[string] and do the concat in there?

your suggestion would require changing `doAssert/assert` API's to allow passing in multiple arguments, but even in this case I don't see how this would help with what this PR is doing, `proc failedAssertImpl*(msg: varargs[string])` would still make argument stringification happen in caller of `failedAssertImpl`, ie, instruction cache bloat would still be there in affected code. And even if all arguments passed were strings already so that `$` is not called, storing those strings would affect instruction cache, see analysis in #14905

Not that there's anything suprising with it, but `proc failedAssertImpl*(msg: varargs[string])` still gives the same bad performance in `example2` in top PR msg. 

> would prefer a solution that doesn't require the use of even more defines to alter the behavior

don't use this if you don't want to, it exists for the same reason as `--stacktrace:on|off` exists, which is that you can't satisfy debuggability and performance simultaneously.

**When running benchmarking or profiling (which translates in deciding whether algo1 or algo 2 is faster), your measurements should not be affected by things that are only there for debuggability.**

> We should do profiling and optimize the code so that it'd perform optimally without degrading the debugging experience.

I did, see analysis in linked issue and in this PR.

**Comment by mratsim** [comment]
>     * new: -d:nimDisableAssertComputedMsgs: miminizes code bloat around the assert calls but can increase binary size; see implementation: it uses a 0-argument function call and discards the runtime message if it can't be computed at CT, but keeps the loc+expr(+ maybe msg if computable at CT); each assert location in source code gives rise to a separate function call.
>       Since it's a separate function call and has 0 arguments, it doesn't affect hotspot much, but it generates a lot of "tiny functions", and may incrIease code size.

I'm in favor of a solution like this but we should avoid adding yet another new flag, it's already hard to track them at the moment. Note: the function must be tagged `{.noinline.}`
Passing the values to display them that function should be possible with minimal overhead as well, if we only convert them to string in the generated error formatter as we can assume that they are already in registers since they are needed for the logical comparison.

**Comment by alaviss** [comment]
> When running benchmarking or profiling (which translates in deciding whether algo1 or algo 2 is faster), your measurements should not be affected by things that are only there for debuggability.

Yes but I think you're missing the actual problem here: These basic debugging primitives (`assert`/`doAssert`) should be (relatively) cheap to begin with. If they produce a significant slowdown aside from what caused by the associated conditional, then that's an issue that we need to solve.

**Comment by timotheecour** [comment]
> I'm in favor of a solution like this but we should avoid adding yet another new flag, it's already hard to track them at the moment.

it's documented (see PR) and when it comes to optimization, you want options.

> Note: the function must be tagged {.noinline.}

I already added {.noinline.} to `failedAssertImpl` in this PR

> Passing the values to display them that function should be possible with minimal overhead as well, if we only convert them to string in the generated error formatter

what you're suggesting would change the assert/doAssert API, which currently takes a msg string argument instead of individual values; it'd be a much more involved change that the simple flag defined in this PR. Furthermore, even then this would have a cost even if you're simply passing those arguments to some hypothetical `doAssertCustom` and won't reduce binary size since you'll end up with 1 generated function per unique argument types. 

> as we can assume that they are already in registers since they are needed for the logical comparison.

that's not true, the context message can include variables that are not held in registers, it gives context to how we got to this error, eg:
```nim
proc fn[T](a: T, file: string) =
  let b = process(a, file)
  for i in b:
    doAssert cond, $(i, a, file) # <= `a` may not be held in register at this point
    doAssertCustom(cond, i, a, file) # even in this case can incur a cost since passed arguments may not be in cache
```

## bottom line:
I don't want to compromise on either debuggability or performance, I want to be able to choose.
* When doing perf comparison/benchmarking, it's much worse to run the risk of jumping to wrong conclusions (as demonstrated in linked issue) than to have a simple additional flag thats documented.
* when not doing comparison/benchmarking, I want my assert failures to show relevant context without worrying about their effect on inlining

**Comment by alaviss** [comment]
> what you're suggesting would change the assert/doAssert API

I think what he meant is that instead of doing this:

```nim
failedAssertImpl(ploc & " `" & expr & "` " & msg)
```

you do this:

```nim
failedAssertImpl(ploc, " `", expr, "` ", msg) # failedAssertImpl(varargs[string])
```

or even this:

```nim
failedAssertImpl(ploc, expr, msg) # failedAssertImpl(string, string, string)
```

This should group all string concat code into `failedAssertImpl`, and we avoid the temporaries associated with them. No `assert`/`doAssert` API change needed at all.

**Comment by timotheecour** [comment]
you're still missing the point: computing `msg` itself can add code bloat affecting inlining and performance. `msg` is a template parameter typed as a string, but can involve runtime computation and in fact this is a common use case when you're showing runtime context on assert failure. This PR allows to optimize it out, and, with `-d:nimDisableAssertComputedMsgs`, only optimizes it out if it can't be computed at CT.

## example
in example given in top post, simply replace `doAssert result != -3, $(result, i, n)`
by
`doAssert result != -3, "result: " & $result`
(which is commonly used in assert messages)

now, even if I omit ploc and expr and simply call `failedAssertImpl(msg)`, the `msg` doesn't get optimized out and results in the **same bad performance (11x slower than with `-d:nimDisableAssertComputedMsgs` and `-d:nimDisableAssertMsgs` )**

even if you optimized how `"result: " & $result` is computed, there will always be cases like this where the assert msg involves computation that affects inlining, even if the assert is not triggered the performance can be affected as demonstrated here.

furthermore even when `msg` is not specified, `failedAssertImpl(ploc, expr, msg)` is still subject to the problem mentioned in https://github.com/nim-lang/Nim/issues/14905 where inlining can be affected by irrelevant changes such as adding a newline which affects `ploc`

which is why I'm introducing both `-d:nimDisableAssertComputedMsgs` and `-d:nimDisableAssertMsgs`

**Comment by arnetheduck** [comment]
meh - the exception raising also adds bloat - any time you use `doAssert` you add bloat and this bloat will have effect on your real-world performance, as will the inlining parameters, the optimization flags that you're passing to the compiler - or the version of the c compiler itself.

If the compiler chooses different inlining strategies for microbenchmarks that have little do do with reality - based on some trivial differences, what do you think it will do for an actual, real application?

If you're using doAssert in your microbenchmark and relying on to guide your performance work, you're solving the wrong problem here - you should consider what it is you're benchmarking really - these flags look inelegant and add bloat to the already existing excessive `when` jungle in `system.nim`, creating yet another dialect of nim.

In fact, it seems more probable that you should thank the compiler for highlighting the potentially unstable performance issues int your codebase.

**Comment by Araq** [comment]
We need a new assert, one that does not take a ``msg`` and uses the instantiation line info because that's a simple ``cstring`` that won't be constructed at runtime.

**Comment by timotheecour** [comment]
@araq reviving this in light of `--lean` proposed for https://github.com/nim-lang/Nim/pull/14282 in https://github.com/nim-lang/Nim/pull/14282#issuecomment-745144807 which is very similar.

The idea is that `--lean` can be queried by user code / stdlib (via `compileOption("lean")`) to produce leaner messages, thus primarily reducing binary code size, but also in some cases also useful for benchmarking since it reduces variability as demonstrated in this PR and https://github.com/nim-lang/Nim/issues/14905.

> We need a new assert, one that does not take a msg and uses the instantiation line info because that's a simple cstring that won't be constructed at runtime.

this PR enables exactly that but in a better way:
* without the major downside of requiring users to patch their code/choose between `s/assert/assertWithoutMsg/`; instead a simple flag `nim --lean [flags...] main` will do.
* msg is still honored so long it's known at CT, otherwise it's dropped: see the PR
```nim
const msg2: cstring = ploc & " `" & expr & "` " & (when msg is static: msg else: "<ommitted>")
```
* it's a trivial diff from this PR to use a cstring instead of a string, but this likely makes no difference since GC is never invoked either way: with string you get:
```c
STRING_LITERAL(TM__xxx, "t11513.nim(8, 14) `rand(int) != 12345` ", 88);
N_LIB_PRIVATE N_NOINLINE(void, colonanonymous___xxx)(void) {
	failedAssertImpl__xxx(((NimStringDesc*) &TM__xxx));
}
```
with cstring you get:
```c
N_LIB_PRIVATE N_NOINLINE(void, colonanonymous___xxx)(void) {
  failedAssertImpl__xxx("t11513.nim(8, 14) `rand(int) != 12345` ");
}
```
